tools/kprint/addBrackets.cpp

```diff
@@ -0,0 +1,359 @@
+ *          BARE_LEFT if there is a terminal only on the right of this nonterminal
+ *          BARE_LEFT if there is a terminal only on the left of this nonterminal
+ *          BARE_BOTH if there are not terminals on either side of this nonterminal
+ */
+Fixity getFixity(int position, KORESymbol *sym, PrettyPrintData const& data) {
+  int result = EMPTY;
+  std::string terminals = data.terminals.at(sym->getName());
+  int terminalPos = getNTPositionInProd(terminals, position);
+  if (!hasTerminalAtIdx(terminals, terminalPos+1)) {
+    result |= BARE_RIGHT;
+  }
+  if (!hasTerminalAtIdx(terminals, terminalPos-1)) {
+    result |= BARE_LEFT;
+  }
+  return (Fixity) result;
+}
+
+/**
+ * Compute whether there are terminals on the far left and far right of a
+ * production.
+ *
+ * sym: the symbol to examine
+ *
+ * data: pretty printing metadata
+ *
+ * returns: EMPTY if there are terminals on both edges of this production.
+ *          BARE_LEFT if there is a terminal only on the right edge of this production
+ *          BARE_RIGHT if there is a terminal only on the left edge of this production
+ *          BARE_BOTH if there are not terminals on either edge of this production
+ */
+Fixity getFixity(KORESymbol *sym, PrettyPrintData const& data) {
+  std::string name = sym->getName();
+  int result = EMPTY;
+  if (data.terminals.at(name)[0] == '0') {
+    result = result | BARE_LEFT;
+  }
+  if (data.terminals.at(name)[data.terminals.at(name).size()-1] == '0') {
+    result = result | BARE_RIGHT;
+  }
+  return (Fixity)result;
+}
+
+/**
+ * Compute the left capture of a pattern.
+ *
+ * If you were to unparse the AST of a term without any parentheses as a
+ * sequence of terminals, start from the first terminal that is part of a
+ * particular pattern (either directly or nested) and proceed left until you
+ * reach the first terminal that is part of a pattern (directly) that is a
+ * parent of that pattern, the pattern whose terminal you reach would be the
+ * left capture of that pattern. Not all terms have a left capture; for
+ * example, the left capture of the top term in an AST does not exist.
+ *
+ * previousLeftCapture: the left capture of `outer`
+ *
+ * outer: the parent of the symbol you wish to compute the left capture of
+ *
+ * position: the position within the nonterminals of `outer` of the pattern
+ * you wish to compute the left capture of.
+ *
+ * data: pretty printing metadata
+ *
+ * returns: the left capture of the term, or NULL if no such term exists
+ */ 
+KORECompositePattern *getLeftCapture(KORECompositePattern *previousLeftCapture, KORECompositePattern *outer, int position, PrettyPrintData const& data) {
+  Fixity fixity = getFixity(outer->getConstructor(), data);
+  if (position == 0 && (fixity & BARE_LEFT)) {
+    return previousLeftCapture;
+  } else {
+    return outer;
+  }
+}
+
+/**
+ * Compute the right capture of a pattern.
+ *
+ * If you were to unparse the AST of a term without any parentheses as a
```

Comment:
I coudn't find where we check that "P" below has a terminal. Other than that, would the following text be more clear?

 * If you were to unparse the AST of a term without any parentheses as a
 * sequence of terminals, start from the point where a particular pattern P
 * (the one at "position" in outer) ends and proceed right until you
 * reach the first terminal that is part of a pattern Q (directly) such that
 * Q is a parent of P. Then Q would be the right capture of that pattern.
 * Not all terms have a right capture; for example, the right capture of the
 * top term in an AST does not exist.

Also, do we always have terminals between non-terminals? If not, then this description is likely wrong.